cannon/mipsevm/exec/mips_instructions.go

```diff
@@ -245,45 +324,107 @@ func ExecuteMipsInstruction(insn, opcode, fun, rs, rt, mem uint32) uint32 {
 				for ; rs&0x80000000 != 0; i++ {
 					rs <<= 1
 				}
-				return i
+				return Word(i)
 			}
 		case 0x0F: // lui
-			return rt << 16
+			return SignExtend(rt<<16, 32)
 		case 0x20: // lb
-			return SignExtend((mem>>(24-(rs&3)*8))&0xFF, 8)
+			msb := uint32(arch.WordSize - 8) // 24 for 32-bit and 56 for 64-bit
+			return SignExtend((mem>>(msb-uint32(rs&arch.ExtMask)*8))&0xFF, 8)
 		case 0x21: // lh
-			return SignExtend((mem>>(16-(rs&2)*8))&0xFFFF, 16)
+			msb := uint32(arch.WordSize - 16) // 16 for 32-bit and 48 for 64-bit
+			mask := Word(arch.ExtMask - 1)
+			return SignExtend((mem>>(msb-uint32(rs&mask)*8))&0xFFFF, 16)
 		case 0x22: // lwl
 			val := mem << ((rs & 3) * 8)
-			mask := uint32(0xFFFFFFFF) << ((rs & 3) * 8)
-			return (rt & ^mask) | val
+			mask := Word(uint32(0xFFFFFFFF) << ((rs & 3) * 8))
+			return SignExtend(((rt & ^mask)|val)&0xFFFFFFFF, 32)
 		case 0x23: // lw
+			// TODO: port to MIPS64
```

Comment:
Probably worth referencing an issue here (could just be a general support 64bit issue). That way if we close the issue without fixing this we'll get failures to tell us we missed it.